### DISCUSSION TITLE
Feature proposal - database access configuration in shell command parameters

### DISCUSSION THREAD

**Discussion by isfag** [AayushK47/fireshell#7] [Ideas]
Hi,

This package is useful for me to make tests on instance of Cloud Firestore Database.
I need almost to use this kind of tool but without prompt interaction, and directly specifying database access in parameters of the batch command.

Do you think in may possible to implement a new version with this functionnality ?

I think it should be useful for all.

Thanks

**Comment by AayushK47** [comment]
HI,

Thanks for the suggestion. Yes, it can be done. I am thinking of adding a config file where you can specify all the parameters required. If the config file is found, you won't get any prompts and directly start querying. That sound better for your use case??

**Comment by AayushK47** [comment]
But before starting on this I found some major bugs on this app due to dependency change that I need to fix first.

Thanks for the suggestion though.

**Comment by isfag** [comment]
Hi @AayushK47,

Thanks a lot for your answer.

I like your following idea : " I am thinking of adding a config file where you can specify all the parameters required. If the config file is found, you won't get any prompts and directly start querying"

I thougth something like the possibility to directly write the following kind of command line :

Fireshell -a [Path To Firebase / Firestore API KEY JSON file] -q [QUERY]

I see the project id is present in the Firebase / Firestore JSON File, so maybe is there any way to simply the configuration using only the JSON file for both Firebase and Firestore ?

Thanks

**Comment by AayushK47** [comment]
Hi @isfag,

I have already published a new release on npm after adding the feature that I mentioned above. It's not perfect yet as I will have to add a feature to reset the config and also a feature add multiple configs. But I hope the release is good enough for your work right now as you won't have to answer the prompts again and again. Just answer the prompt 1st time you run the shell. After that, it will connect with the previously provided answers.

**Comment by AayushK47** [comment]
The possibility you mentioned in your last comment might not be as good as you think. Firstly, to implement what you are saying, I'll have to switch a package from inquirer to commander. Secondly, It will be a single CLI command that the user will have to enter repeatedly to execute 1 query. I think the current implementation is better since, once the shell receives all answers, you can run queries directly. Thirdly, if the project goes that way, we won't be able to implement a feature I have been thinking of - multiline inputs.

Thanks anyway for your suggestion. If you have any other feature ideas do share, your feedback will be very valuable.

Thanks.

**Comment by isfag** [comment]
Thanks @AayushK47. I understand.

In my use cases, I would like to use command line to admin or to test the database in industrial way.
Insert several times the same parameter at each line is not really an inconvenient for me.

Let me argument my proposal :

If we can directly use your tool in a one command line without prompt, like followed:
fireshell -a [PathToJSON] -q [QUERY]

We could then save our works. 
If you have several command lines, you could save your work for exemple like followed: 

 firefeshell -a [PathToJSON Database A] -q [QUERY A]
 firefeshell -a [PathToJSON Database A] -q [QUERY B]
 firefeshell -a [PathToJSON Database A] -q [QUERY C]
 firefeshell -a [PathToJSON Database A] -q [QUERY D]

 firefeshell -a [PathToJSON Database A] -q [QUERY A]
 firefeshell -a [PathToJSON Database B] -q [QUERY B]

Then we could re-used this work in simply way in a batch programme, or by copy-past in our CLI.

While currently, if I understand well, you should firstly pass by prompt and then, the query command line is inside a prompt mode.
I my suggestion, the query command line should pass, from the prompt mode, to the main command line mode.

I understand if my suggestion not convainced you, and if that this one may implicate a major modification.

In all cases thank you for your listening.

**Reply by AayushK47** [reply]
Ohk, now I have a better picture of what your require. What I can do is add a feature like mongoimport. You can write all your queries in 1 file and pass that file to the cli. The cli will run all the commands and return the output in a file. Or it could be more generalized.

Thanks a lot.

**Comment by isfag** [comment]
And thanks for the new version ;-)

**Comment by isfag** [comment]
Yes! After reading mongoimport documentation, it seem to be a good comparaison.
And it show me that an output file parameter for the "GET" query type should be most usefull again.

So with this add, it become something like followed (if the QUERY type is GET):

fireshell -a [PathToJSON] -q [QUERY GET] -o [OUPTUT FILE] -f [OUTPUT FILE FORMAT]

But I imagine it should implicate a very big build.

In my research, I don't find an existing open source tool like this again for firestore, and it seem also not again possible with the official firestore CLI tool.

I think that if your tool is increased in command line usable in batch program, it will be easiest to integrate it for more use cases.

Thanks,

**Reply by AayushK47** [reply]
I'll see what can be done.

Thanks for all your suggestions.

**Reply by AayushK47** [reply]
BTW, If you find this project useful, please do star the repo.

**Comment by isfag** [comment]
Okey, let me know if you make something about this feature.

Thanks,

**Comment by AayushK47** [comment]
Hey there @isfag,

I tried adding the feature you wanted but with no success. Apperently, node starts a shell when you end the cli command with a parenthesis.

Nonetheless I have released a new version of fireshell using which you can run multiple queries at once. Checkout the readme on [here](https://github.com/AayushK47/fireshell/) or in [npmjs](https://www.npmjs.com/package/fireshell).
